src/Version.h

```diff
@@ -17,26 +17,79 @@
+const std::uint64_t FIELD_PATCH   = (FIELD_WIDTH << Version::OFFSET_PATCH);
+
+using exact_width_mumble_raw_version_t = std::uint64_t;
+using mumble_raw_version_t             = unsigned long int;
 
 static_assert(sizeof(exact_width_mumble_raw_version_t) <= sizeof(mumble_raw_version_t),
 			  "Invalid scalar type used to represent Mumble versions");
 
 constexpr mumble_raw_version_t UNKNOWN = 0;
 
-unsigned int getRaw(const QString &version = QLatin1String(MUMTEXT(MUMBLE_VERSION)));
+mumble_raw_version_t getRaw(const QString &version = QLatin1String(MUMTEXT(MUMBLE_VERSION)));
 QString toString(mumble_raw_version_t version);
 bool get(int *major, int *minor, int *patch, const QString &version = QLatin1String(MUMTEXT(MUMBLE_VERSION)));
 
-constexpr Version::mumble_raw_version_t toRaw(int major, int minor, int patch) {
-	return (major << 16) | (minor << 8) | patch;
+constexpr std::uint16_t getMajor(Version::mumble_raw_version_t version) {
+	return static_cast< std::uint16_t >((version & Version::FIELD_MAJOR) >> Version::OFFSET_MAJOR);
+}
+
+constexpr std::uint16_t getMinor(Version::mumble_raw_version_t version) {
+	return static_cast< std::uint16_t >((version & Version::FIELD_MINOR) >> Version::OFFSET_MINOR);
+}
+
+constexpr std::uint16_t getPatch(Version::mumble_raw_version_t version) {
+	return static_cast< std::uint16_t >((version & Version::FIELD_PATCH) >> Version::OFFSET_PATCH);
+}
+
+constexpr Version::mumble_raw_version_t toRaw(std::uint16_t major, std::uint16_t minor, std::uint16_t patch) {
+	return ((static_cast< std::uint64_t >(major) << Version::OFFSET_MAJOR)
+			| (static_cast< std::uint64_t >(minor) << Version::OFFSET_MINOR)
+			| (static_cast< std::uint64_t >(patch) << Version::OFFSET_PATCH));
+}
+
+constexpr void fromRaw(Version::mumble_raw_version_t version, int *major, int *minor, int *patch) {
+	*major = Version::getMajor(version);
+	*minor = Version::getMinor(version);
+	*patch = Version::getPatch(version);
+}
+
+//
+// Legacy versions: These versions are kept around for backward compatibility, but
+// have since been replaced by other version formats.
+//
+// Mumble legacy version format 1 (uint32):
+// major   minor  patch
+// 0xFFFF  0xFF   0xFF
+// (big-endian)
+//
+
+constexpr Version::mumble_raw_version_t fromLegacyVersion(std::uint32_t version) {
+	return toRaw((version & 0xFFFF0000) >> 16, (version & 0xFF00) >> 8, version & 0xFF);
 }
 
-constexpr void fromRaw(unsigned int version, int *major, int *minor, int *patch) {
-	*major = (version & 0xFFFF0000) >> 16;
-	*minor = (version & 0xFF00) >> 8;
-	*patch = (version & 0xFF);
+constexpr std::uint32_t toLegacyVersion(Version::mumble_raw_version_t version) {
+	return ((std::min(static_cast< uint32_t >(Version::getMajor(version)), static_cast< uint32_t >(0xFFFF)) << 16)
+			| (std::min(static_cast< uint32_t >(Version::getMinor(version)), static_cast< uint32_t >(0xFF)) << 8)
+			| std::min(static_cast< uint32_t >(Version::getPatch(version)), static_cast< uint32_t >(0xFF)));
```

Comment:
I think I would prefer expressing the max-values via e.g. `std::numeric_limits< std::uint16_t>::max`. Then these constants don't quite feel like "magic numbers" :)